stdlib/lang-parser/parserModule/parser.bal

```diff
@@ -0,0 +1,1260 @@
+            return valKind;
+        } else {
+            return ERROR_VALUE_TYPE;
+        }
+    }
+
+    //check the operator kind of the operator token
+    function matchOperatorType(Token operator) returns OperatorKind {
+        var opKind = self.oprMap.operatorTypeMap[tokenNames[operator.tokenType]];
+
+        if (opKind is OperatorKind) {
+            return opKind;
+        } else {
+            return ERROR_OP;
+        }
+    }
+
+};
+
+#operator stack to store the operators
+type OperatorStack object {
+    Token[] opStack = [];
+    int top;
+
+    public function __init(int top = 0) {
+        self.top = top;
+    }
+
+    function push(Token token2) {
+        self.opStack[self.top] = token2;
+        self.top = self.top + 1;
+    }
+
+    function pop() returns Token {
+        if (self.top != 0) {
+            self.top = self.top - 1;
+            Token operatorToken = self.opStack[self.top];
+            return operatorToken;
+        } else {
+            return {
+                tokenType: PARSER_ERROR_TOKEN,
+                text: "<missing operator>",
+                startPos: -1,
+                endPos: -1,
+                lineNumber: 0,
+                index: -1,
+                whiteSpace: ""
+            };
+        }
+    }
+
+    function peek() returns int {
+        int topTkn2 = -1;
+        if (self.top != 0) {
+            Token topTkn = self.opStack[self.top - 1];
+            topTkn2 = topTkn.tokenType;
+        }
+        return topTkn2;
+    }
+
+    function isEmpty() returns boolean {
+        if (self.opStack.length() == 1) {
+            return false;
+        }
+        return true;
+    }
+
+    function size() returns int {
+        int opSize = self.opStack.length();
+        return opSize;
+    }
+
+    function opPrecedence(int opToken) returns int {
+        if (opToken == COMMA) {
+            return 0;
+        } else if (opToken == REF_EQUAL || opToken == REF_NOT_EQUAL) {
+            return 2;
+        } else if (opToken == EQUAL || opToken == NOT_EQUAL) {
+            return 3;
+        } else if (opToken == LT_EQUAL || opToken == GT_EQUAL || opToken == LT || opToken == GT) {
+            return 4;
+        } else if (opToken == ADD || opToken == SUB) {
+            return 5;
+        } else if (opToken == DIV || opToken == MUL || opToken == MOD) {
+            return 6;
+        } else if (opToken == UNARY_MINUS || opToken == UNARY_PLUS || opToken == NOT || opToken == BIT_COMPLEMENT || opToken == UNTAINT) {
+            return 7;
+        }
+        return -1;
+    }
+};
+
+# Expression stack stores each expression built
+type ExprStack object {
+    ExpressionNode[] exprStack = [()];
+    int top;
+
+    public function __init(int top = 0) {
+        self.top = top;
+    }
+
+    function push(ExpressionNode token2) {
+        self.exprStack[self.top] = token2;
+        self.top = self.top + 1;
+    }
+
+    function pop() returns ExpressionNode {
+        if (self.top != 0) {
```

Comment:
I'm assuming `ExpressionNode` has `()` as a value, which is why this works?

If so, shall we change this to
```ballerina
        if (self.top == 0) {
            return;
        }
        self.top = self.top - 1;
        return self.exprStack[self.top];
```